File: Algorithms/DLX/Zig/dlx.zig
const std = @import("std");

// DLX node structure
const DlxNode = struct {
    left: ?*DlxNode,
    right: ?*DlxNode,
    up: ?*DlxNode,
    down: ?*DlxNode,
    column: ?*DlxNode,
    size: usize,
    row_id: i32,
    col_id: i32,
};

var dlx_iterations: usize = 0;

fn coverColumn(c: *DlxNode) void {
    // Remove column from header list
    c.right.?.*.left = c.left;
    c.left.?.*.right = c.right;

    // For each row in this column
    var row_node = c.down;
    while (row_node != null and row_node.? != c) {
        // For each node in this row
        var right_node = row_node.?.*.right;
        while (right_node != null and right_node.? != row_node.?) {
            // Remove from column
            right_node.?.*.down.?.*.up = right_node.?.*.up;
            right_node.?.*.up.?.*.down = right_node.?.*.down;
            right_node.?.*.column.?.*.size -= 1;
            right_node = right_node.?.*.right;
        }
        row_node = row_node.?.*.down;
    }
}

fn uncoverColumn(c: *DlxNode) void {
    // For each row in this column (in reverse order)
    var row_node = c.up;
    while (row_node != null and row_node.? != c) {
        // For each node in this row (in reverse order)
        var left_node = row_node.?.*.left;
        while (left_node != null and left_node.? != row_node.?) {
            // Restore to column
            left_node.?.*.column.?.*.size += 1;
            left_node.?.*.down.?.*.up = left_node;
            left_node.?.*.up.?.*.down = left_node;
            left_node = left_node.?.*.left;
        }
        row_node = row_node.?.*.up;
    }

    // Restore column to header list
    c.right.?.*.left = c;
    c.left.?.*.right = c;
}

fn chooseColumn(root: *DlxNode) ?*DlxNode {
    var best: ?*DlxNode = null;
    var min_size: usize = std.math.maxInt(usize);

    var col_node = root.right;
    while (col_node != null and col_node.? != root) {
        if (col_node.?.*.size < min_size) {
            min_size = col_node.?.*.size;
            best = col_node;
        }
        col_node = col_node.?.*.right;
    }

    return best;
}

fn dlxSearch(root: *DlxNode, k: usize, solution: []i32) bool {
    dlx_iterations += 1;

    // If matrix is empty, we found a solution
    if (root.right == null or root.right.? == root) {
        return true;
    }

    // Choose column with minimum size
    const col = chooseColumn(root) orelse return false;

    // If column has no rows, no solution possible
    if (col.size == 0) {
        return false;
    }

    // Cover this column
    coverColumn(col);

    // Try each row in this column
    var row_node = col.down;
    while (row_node != null and row_node.? != col) {
        // Add row to partial solution
        solution[k] = row_node.?.*.row_id;

        // Cover all other columns in this row
        var right_node = row_node.?.*.right;
        while (right_node != null and right_node.? != row_node.?) {
            if (right_node.?.*.column) |col_to_cover| {
                coverColumn(col_to_cover);
            }
            right_node = right_node.?.*.right;
        }

        // Recurse
        if (dlxSearch(root, k + 1, solution)) {
            return true;
        }

        // Backtrack: uncover all columns in this row
        var left_node = row_node.?.*.left;
        while (left_node != null and left_node.? != row_node.?) {
            if (left_node.?.*.column) |col_to_uncover| {
                uncoverColumn(col_to_uncover);
            }
            left_node = left_node.?.*.left;
        }

        row_node = row_node.?.*.down;
    }

    // Uncover column
    uncoverColumn(col);

    return false;
}

fn createNode(allocator: std.mem.Allocator) !*DlxNode {
    const node = try allocator.create(DlxNode);
    node.* = DlxNode{
        .left = null,
        .right = null,
        .up = null,
        .down = null,
        .column = null,
        .size = 0,
        .row_id = -1,
        .col_id = -1,
    };
    return node;
}

const MatrixResult = struct {
    root: *DlxNode,
    row_info: [729]RowInfo,
    row_starts: [729]?*DlxNode,
};

const RowInfo = struct {
    row: i32,
    col: i32,
    num: i32,
};

fn buildSudokuMatrix(allocator: std.mem.Allocator, puzzle: [9][9]i32) !MatrixResult {
    // Create root node
    const root = try createNode(allocator);
    root.left = root;
    root.right = root;
    root.up = root;
    root.down = root;

    // Create column headers (324 constraints: 81 cell + 81 row + 81 col + 81 box)
    var columns: [324]*DlxNode = undefined;
    for (0..324) |i| {
        const col = try createNode(allocator);
        col.size = 0;
        col.col_id = @as(i32, @intCast(i));
        col.up = col;
        col.down = col;
        col.column = col;

        // Add to header list
        col.left = root.left;
        col.right = root;
        root.left.?.*.right = col;
        root.left = col;

        columns[i] = col;
    }

    // Track row metadata
    var row_info: [729]RowInfo = undefined;
    var row_starts: [729]?*DlxNode = undefined;
    for (0..729) |i| {
        row_starts[i] = null;
    }
    var row_id: usize = 0;

    // Add rows to the matrix
    for (0..9) |row| {
        for (0..9) |col| {
            const val = puzzle[row][col];
            if (val == 0) {
                // Empty cell: add rows for all possible digits
                for (1..10) |digit| {
                    row_starts[row_id] = try addRow(allocator, &columns, &row_info, row_id, @as(i32, @intCast(row)), @as(i32, @intCast(col)), @as(i32, @intCast(digit)));
                    row_id += 1;
                }
            } else {
                // Fixed cell: add only one row
                row_starts[row_id] = try addRow(allocator, &columns, &row_info, row_id, @as(i32, @intCast(row)), @as(i32, @intCast(col)), val);
                row_id += 1;
            }
        }
    }

    return MatrixResult{
        .root = root,
        .row_info = row_info,
        .row_starts = row_starts,
    };
}

fn addRow(allocator: std.mem.Allocator, columns: *[324]*DlxNode, row_info: *[729]RowInfo, row_id: usize, row: i32, col: i32, digit: i32) !*DlxNode {
    // Store row metadata
    row_info[row_id] = RowInfo{
        .row = row,
        .col = col,
        .num = digit,
    };

    // Calculate constraint column indices
    const cell_idx = row * 9 + col;
    const row_idx = 81 + row * 9 + (digit - 1);
    const col_idx = 162 + col * 9 + (digit - 1);
    const box_idx = 243 + ((@divFloor(row, 3) * 3 + @divFloor(col, 3)) * 9 + (digit - 1));

    const indices = [4]usize{
        @as(usize, @intCast(cell_idx)),
        @as(usize, @intCast(row_idx)),
        @as(usize, @intCast(col_idx)),
        @as(usize, @intCast(box_idx)),
    };

    var first_node: ?*DlxNode = null;
    var prev_node: ?*DlxNode = null;

    // Create nodes for each constraint
    for (indices) |idx| {
        const node = try createNode(allocator);
        node.row_id = @as(i32, @intCast(row_id));
        node.col_id = @as(i32, @intCast(idx));
        node.column = columns[idx];

        // Add to column list
        node.up = columns[idx].up;
        node.down = columns[idx];
        columns[idx].up.?.*.down = node;
        columns[idx].up = node;
        columns[idx].size += 1;

        // Link horizontally
        if (first_node == null) {
            first_node = node;
            node.left = node;
            node.right = node;
        } else {
            node.left = prev_node;
            node.right = first_node;
            prev_node.?.*.right = node;
            first_node.?.*.left = node;
        }
        prev_node = node;
    }

    return first_node.?;
}

fn coverClues(puzzle: [9][9]i32, row_info: [729]RowInfo, row_starts: [729]?*DlxNode) void {
    for (0..9) |r| {
        for (0..9) |c| {
            if (puzzle[r][c] != 0) {
                const n = puzzle[r][c];
                // Find the row for this clue
                for (0..729) |row_id| {
                    if (row_starts[row_id]) |start_node| {
                        if (row_info[row_id].row == @as(i32, @intCast(r)) and
                            row_info[row_id].col == @as(i32, @intCast(c)) and
                            row_info[row_id].num == n)
                        {
                            // Cover all columns in this row
                            var curr = start_node;
                            var first = true;
                            while (first or curr != start_node) {
                                first = false;
                                if (curr.column) |col| {
                                    coverColumn(col);
                                }
                                curr = curr.right.?;
                            }
                            break;
                        }
                    }
                }
            }
        }
    }
}

fn parseMatrixFile(allocator: std.mem.Allocator, filename: []const u8) ![9][9]i32 {
    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var puzzle: [9][9]i32 = undefined;
    var row_idx: usize = 0;

    var lines = std.mem.tokenizeScalar(u8, content, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;
        if (line[0] < '0' or line[0] > '9') continue;

        var col_idx: usize = 0;
        var it = std.mem.tokenizeAny(u8, line, " \r\t");
        while (it.next()) |token| {
            if (col_idx < 9) {
                puzzle[row_idx][col_idx] = try std.fmt.parseInt(i32, token, 10);
                col_idx += 1;
            }
        }
        if (col_idx > 0) {
            row_idx += 1;
        }
        if (row_idx >= 9) break;
    }

    return puzzle;
}

fn extractSolution(solution: []i32, k: usize, puzzle: [9][9]i32, row_info: [729]RowInfo) [9][9]i32 {
    var grid = puzzle; // Start with the original puzzle (includes clues)

    for (0..k) |i| {
        const row_id = solution[i];
        if (row_id >= 0 and row_id < 729) {
            const info = row_info[@as(usize, @intCast(row_id))];
            grid[@as(usize, @intCast(info.row))][@as(usize, @intCast(info.col))] = info.num;
        }
    }

    return grid;
}

fn printToStdout(comptime fmt: []const u8, args: anytype) void {
    var buf: [4096]u8 = undefined;
    const slice = std.fmt.bufPrint(&buf, fmt, args) catch return;
    _ = std.posix.write(1, slice) catch {};
}

fn printGrid(grid: [9][9]i32) void {
    for (grid) |row| {
        for (row, 0..) |cell, i| {
            printToStdout("{d}", .{cell});
            if (i < 8) printToStdout(" ", .{});
        }
        printToStdout("\n", .{});
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    defer _ = gpa.deinit();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len < 2) {
        printToStdout("Usage: {s} <matrix_file>\n", .{args[0]});
        return;
    }

    // Parse input puzzle
    const puzzle = try parseMatrixFile(allocator, args[1]);

    // Build DLX matrix
    const matrix = try buildSudokuMatrix(allocator, puzzle);

    // Cover pre-filled clues
    coverClues(puzzle, matrix.row_info, matrix.row_starts);

    // Solve
    var solution: [81]i32 = undefined;
    dlx_iterations = 0;

    if (dlxSearch(matrix.root, 0, &solution)) {
        const solved = extractSolution(&solution, 81, puzzle, matrix.row_info);
        printGrid(solved);
        printToStdout("Solved in Iterations={d}\n", .{dlx_iterations});
    } else {
        printToStdout("No solution found.\n", .{});
    }
}

File: Algorithms/DLX/Erlang/dlx.erl
#!/usr/bin/env escript
%% -*- erlang -*-
%% Erlang Dancing Links (DLX) Sudoku Solver
%% Port of C DLX implementation using ETS table for mutable state

main([Filename|_]) ->
    %% Initialize iteration counter
    ets:new(state, [set, named_table, public]),
    ets:insert(state, {dlx_iterations, 0}),
    ets:insert(state, {node_count, 0}),

    %% Read and print puzzle
    Puzzle = read_matrix(Filename),
    print_puzzle("Puzzle", Puzzle),

    %% Solve using DLX
    case solve_dlx(Puzzle) of
        {ok, Solution} ->
            print_puzzle("Puzzle", Solution),
            [{dlx_iterations, Iters}] = ets:lookup(state, dlx_iterations),
            io:format("~nSolved in Iterations=~B~n~n", [Iters]);
        error ->
            io:format("~nNo solution found~n")
    end,

    ets:delete(state);
main(_) ->
    io:format(standard_error, "Usage: escript dlx.erl <matrix_file>~n", []),
    halt(1).

%% Read matrix from file
read_matrix(Filename) ->
    %% Print filename (normalize /app/Matrices to ../Matrices)
    case string:prefix(Filename, "/app/Matrices/") of
        nomatch -> io:format("~s~n", [Filename]);
        Rest -> io:format("../Matrices/~s~n", [Rest])
    end,
    {ok, Binary} = file:read_file(Filename),
    Lines = string:split(binary_to_list(Binary), "\n", all),
    Grid = array:new([{size, 9}, {default, array:new([{size, 9}, {default, 0}])}]),
    read_rows(Lines, 0, Grid).

read_rows(_, 9, Grid) -> Grid;
read_rows([], _, Grid) -> Grid;
read_rows([Line|Rest], Row, Grid) ->
    Trimmed = string:trim(Line),
    case Trimmed of
        "" -> read_rows(Rest, Row, Grid);
        [$#|_] -> read_rows(Rest, Row, Grid);  %% Skip comments
        _ ->
            Numbers = string:tokens(Trimmed, " "),
            NewGrid = parse_row(Numbers, Row, 0, Grid),
            io:format("~n"),
            read_rows(Rest, Row + 1, NewGrid)
    end.

parse_row([], _, _, Grid) -> Grid;
parse_row([NumStr|Rest], Row, Col, Grid) when Col < 9 ->
    Num = list_to_integer(NumStr),
    io:format("~B ", [Num]),
    RowArray = array:get(Row, Grid),
    NewRowArray = array:set(Col, Num, RowArray),
    NewGrid = array:set(Row, NewRowArray, Grid),
    parse_row(Rest, Row, Col + 1, NewGrid);
parse_row(_, _, _, Grid) -> Grid.

%% Print puzzle
print_puzzle(Label, Grid) ->
    io:format("~n~s:~n", [Label]),
    array:map(fun(_R, RowArray) ->
        array:map(fun(_C, Val) ->
            io:format("~B ", [Val]),
            Val
        end, RowArray),
        io:format("~n"),
        RowArray
    end, Grid),
    Grid.

%% Node structure stored in ETS
%% Each node is {NodeId, {left, right, up, down, column, size, row_id, col_id}}

get_node_field(NodeId, Field) ->
    case ets:lookup(state, NodeId) of
        [{NodeId, NodeData}] -> element(Field, NodeData);
        [] -> undefined
    end.

set_node_field(NodeId, Field, Value) ->
    case ets:lookup(state, NodeId) of
        [{NodeId, NodeData}] ->
            NewData = setelement(Field, NodeData, Value),
            ets:insert(state, {NodeId, NewData});
        [] -> ok
    end.

get_node(NodeId) ->
    case ets:lookup(state, NodeId) of
        [{NodeId, NodeData}] -> NodeData;
        [] -> undefined
    end.

set_node(NodeId, NodeData) ->
    ets:insert(state, {NodeId, NodeData}).

%% NodeData = {Left, Right, Up, Down, Column, Size, RowId, ColId}
make_node(NodeId, RowId, ColId) ->
    {NodeId, NodeId, NodeId, NodeId, NodeId, 0, RowId, ColId}.

%% Calculate constraint column indices
get_position_col(R, C) -> {col, R * 9 + C}.
get_row_col(R, N) -> {col, 81 + R * 9 + (N - 1)}.
get_col_col(C, N) -> {col, 162 + C * 9 + (N - 1)}.
get_box_col(R, C, N) ->
    Box = (R div 3) * 3 + (C div 3),
    {col, 243 + Box * 9 + (N - 1)}.

%% Initialize DLX matrix
init_dlx_matrix() ->
    %% Create root column
    set_node(root, make_node(root, -1, -1)),

    %% Create 324 column headers
    lists:foreach(fun(I) ->
        ColId = {col, I},
        set_node(ColId, make_node(ColId, -1, I)),

        %% Link into header list
        {LeftId, RightId, UpId, DownId, ColId2, Size, RowId, ColId3} = get_node(root),
        set_node_field(ColId, 1, LeftId),  %% left
        set_node_field(ColId, 2, root),    %% right
        set_node_field(LeftId, 2, ColId),
        set_node_field(root, 1, ColId)
    end, lists:seq(0, 323)).

%% Add a node to the DLX matrix
add_node(ColId, RowId) ->
    [{node_count, Count}] = ets:lookup(state, node_count),
    NodeId = {node, Count},
    ets:insert(state, {node_count, Count + 1}),

    %% Create node
    set_node(NodeId, make_node(NodeId, RowId, -1)),
    set_node_field(NodeId, 5, ColId),  %% column

    %% Insert at end of column's circular list
    {_,_, UpId, _, _, _, _, _} = get_node(ColId),
    set_node_field(NodeId, 4, ColId),      %% down = column
    set_node_field(NodeId, 3, UpId),       %% up = column.up
    set_node_field(UpId, 4, NodeId),       %% column.up.down = node
    set_node_field(ColId, 3, NodeId),      %% column.up = node

    %% Increment column size
    {_, _, _, _, _, Size, _, _} = get_node(ColId),
    set_node_field(ColId, 6, Size + 1),

    NodeId.

%% Build a DLX row
build_dlx_row(R, C, N, RowId) ->
    %% Store row metadata
    ets:insert(state, {{row_info, RowId}, {R, C, N}}),

    %% Create nodes for the 4 constraints
    N1 = add_node(get_position_col(R, C), RowId),
    N2 = add_node(get_row_col(R, N), RowId),
    N3 = add_node(get_col_col(C, N), RowId),
    N4 = add_node(get_box_col(R, C, N), RowId),

    %% Link nodes horizontally in circular list
    set_node_field(N1, 2, N2),  %% right
    set_node_field(N2, 2, N3),
    set_node_field(N3, 2, N4),
    set_node_field(N4, 2, N1),

    set_node_field(N1, 1, N4),  %% left
    set_node_field(N2, 1, N1),
    set_node_field(N3, 1, N2),
    set_node_field(N4, 1, N3),

    %% Store first node for this row
    ets:insert(state, {{row_start, RowId}, N1}),
    N1.

%% Build matrix from puzzle
build_dlx_matrix(Puzzle) ->
    init_dlx_matrix(),
    build_matrix_rows(Puzzle, 0, 0, 0).

build_matrix_rows(_, 9, _, _) -> ok;
build_matrix_rows(Puzzle, R, 9, RowId) ->
    build_matrix_rows(Puzzle, R + 1, 0, RowId);
build_matrix_rows(Puzzle, R, C, RowId) ->
    RowArray = array:get(R, Puzzle),
    Val = array:get(C, RowArray),
    NewRowId = if
        Val =/= 0 ->
            build_dlx_row(R, C, Val, RowId),
            RowId + 1;
        true ->
            lists:foldl(fun(N, AccRowId) ->
                build_dlx_row(R, C, N, AccRowId),
                AccRowId + 1
            end, RowId, lists:seq(1, 9))
    end,
    build_matrix_rows(Puzzle, R, C + 1, NewRowId).

%% Cover a column
cover_column(ColId) ->
    %% Remove column header from header list
    {LeftId, RightId, _, _, _, _, _, _} = get_node(ColId),
    set_node_field(RightId, 1, LeftId),
    set_node_field(LeftId, 2, RightId),

    %% For each row in this column
    {_, _, _, DownId, _, _, _, _} = get_node(ColId),
    cover_rows(DownId, ColId).

cover_rows(RowNode, ColId) when RowNode =:= ColId -> ok;
cover_rows(RowNode, ColId) ->
    %% For each node in this row
    {_, RightId, _, _, _, _, _, _} = get_node(RowNode),
    cover_row_nodes(RightId, RowNode),

    {_, _, _, DownId, _, _, _, _} = get_node(RowNode),
    cover_rows(DownId, ColId).

cover_row_nodes(RightNode, RowNode) when RightNode =:= RowNode -> ok;
cover_row_nodes(RightNode, RowNode) ->
    %% Remove this node from its column
    {_, _, UpId, DownId, Column, _, _, _} = get_node(RightNode),
    set_node_field(DownId, 3, UpId),
    set_node_field(UpId, 4, DownId),

    {_, _, _, _, _, Size, _, _} = get_node(Column),
    set_node_field(Column, 6, Size - 1),

    {_, RightId2, _, _, _, _, _, _} = get_node(RightNode),
    cover_row_nodes(RightId2, RowNode).

%% Uncover a column
uncover_column(ColId) ->
    %% For each row in this column (in reverse order)
    {_, _, UpId, _, _, _, _, _} = get_node(ColId),
    uncover_rows(UpId, ColId),

    %% Restore column header to header list
    {LeftId, RightId, _, _, _, _, _, _} = get_node(ColId),
    set_node_field(RightId, 1, ColId),
    set_node_field(LeftId, 2, ColId).

uncover_rows(RowNode, ColId) when RowNode =:= ColId -> ok;
uncover_rows(RowNode, ColId) ->
    %% For each node in this row (in reverse order)
    {LeftId, _, _, _, _, _, _, _} = get_node(RowNode),
    uncover_row_nodes(LeftId, RowNode),

    {_, _, UpId, _, _, _, _, _} = get_node(RowNode),
    uncover_rows(UpId, ColId).

uncover_row_nodes(LeftNode, RowNode) when LeftNode =:= RowNode -> ok;
uncover_row_nodes(LeftNode, RowNode) ->
    %% Restore this node to its column
    {_, _, UpId, DownId, Column, _, _, _} = get_node(LeftNode),
    {_, _, _, _, _, Size, _, _} = get_node(Column),
    set_node_field(Column, 6, Size + 1),
    set_node_field(DownId, 3, LeftNode),
    set_node_field(UpId, 4, LeftNode),

    {LeftId2, _, _, _, _, _, _, _} = get_node(LeftNode),
    uncover_row_nodes(LeftId2, RowNode).

%% Choose column with minimum size
choose_column(Root) ->
    {_, RightId, _, _, _, _, _, _} = get_node(Root),
    choose_column_scan(RightId, Root, none, 9999999).

choose_column_scan(ColNode, Root, Best, _MinSize) when ColNode =:= Root ->
    Best;
choose_column_scan(ColNode, Root, none, _MinSize) ->
    {_, RightId, _, _, _, Size, _, _} = get_node(ColNode),
    choose_column_scan(RightId, Root, ColNode, Size);
choose_column_scan(ColNode, Root, Best, MinSize) ->
    {_, RightId, _, _, _, Size, _, _} = get_node(ColNode),
    {NewBest, NewMinSize} = if
        Size < MinSize -> {ColNode, Size};
        true -> {Best, MinSize}
    end,
    choose_column_scan(RightId, Root, NewBest, NewMinSize).

%% DLX Search
dlx_search(Root, K, Solution) ->
    %% Increment iteration counter
    [{dlx_iterations, Iters}] = ets:lookup(state, dlx_iterations),
    ets:insert(state, {dlx_iterations, Iters + 1}),

    %% If matrix is empty, we found a solution
    {_, RightId, _, _, _, _, _, _} = get_node(Root),
    if
        RightId =:= Root -> {ok, Solution};
        true ->
            %% Choose column with minimum size
            case choose_column(Root) of
                none -> error;
                Col ->
                    {_, _, _, _, _, Size, _, _} = get_node(Col),
                    if
                        Size =:= 0 -> error;
                        true ->
                            cover_column(Col),
                            {_, _, _, DownId, _, _, _, _} = get_node(Col),
                            try_rows(DownId, Col, Root, K, Solution)
                    end
            end
    end.

try_rows(RowNode, Col, _Root, _K, _Solution) when RowNode =:= Col ->
    uncover_column(Col),
    error;
try_rows(RowNode, Col, Root, K, Solution) ->
    %% Add row to partial solution
    {_, _, _, _, _, _, RowId, _} = get_node(RowNode),
    NewSolution = array:set(K, RowId, Solution),

    %% Cover all other columns in this row
    {_, RightId, _, _, _, _, _, _} = get_node(RowNode),
    cover_row_cols(RightId, RowNode),

    %% Recurse
    case dlx_search(Root, K + 1, NewSolution) of
        {ok, FinalSolution} -> {ok, FinalSolution};
        error ->
            %% Backtrack
            {LeftId, _, _, _, _, _, _, _} = get_node(RowNode),
            uncover_row_cols(LeftId, RowNode),

            {_, _, _, DownId, _, _, _, _} = get_node(RowNode),
            try_rows(DownId, Col, Root, K, Solution)
    end.

cover_row_cols(RightNode, RowNode) when RightNode =:= RowNode -> ok;
cover_row_cols(RightNode, RowNode) ->
    {_, _, _, _, Column, _, _, _} = get_node(RightNode),
    cover_column(Column),
    {_, RightId, _, _, _, _, _, _} = get_node(RightNode),
    cover_row_cols(RightId, RowNode).

uncover_row_cols(LeftNode, RowNode) when LeftNode =:= RowNode -> ok;
uncover_row_cols(LeftNode, RowNode) ->
    {_, _, _, _, Column, _, _, _} = get_node(LeftNode),
    uncover_column(Column),
    {LeftId, _, _, _, _, _, _, _} = get_node(LeftNode),
    uncover_row_cols(LeftId, RowNode).

%% Cover given clues
cover_clues(Puzzle) ->
    cover_clues_iter(Puzzle, 0, 0).

cover_clues_iter(_, 9, _) -> ok;
cover_clues_iter(Puzzle, R, 9) ->
    cover_clues_iter(Puzzle, R + 1, 0);
cover_clues_iter(Puzzle, R, C) ->
    RowArray = array:get(R, Puzzle),
    Val = array:get(C, RowArray),
    if
        Val =/= 0 ->
            %% Find the row for this clue by scanning all rows
            find_and_cover_clue_row(R, C, Val);
        true -> ok
    end,
    cover_clues_iter(Puzzle, R, C + 1).

find_and_cover_clue_row(R, C, N) ->
    %% Scan all row_info entries to find the matching row
    AllKeys = ets:match(state, {{row_info, '$1'}, '_'}),
    find_clue_in_rows(lists:flatten(AllKeys), R, C, N).

find_clue_in_rows([], _, _, _) -> ok;
find_clue_in_rows([RowId|Rest], R, C, N) ->
    case ets:lookup(state, {row_info, RowId}) of
        [{{row_info, RowId}, {R, C, N}}] ->
            [{{row_start, RowId}, StartNode}] = ets:lookup(state, {row_start, RowId}),
            cover_clue_cols(StartNode, StartNode);
        _ ->
            find_clue_in_rows(Rest, R, C, N)
    end.

cover_clue_cols(Curr, StartNode) ->
    {_, _, _, _, Column, _, _, _} = get_node(Curr),
    cover_column(Column),
    {_, RightId, _, _, _, _, _, _} = get_node(Curr),
    if
        RightId =:= StartNode -> ok;
        true -> cover_clue_cols(RightId, StartNode)
    end.

%% Solve DLX
solve_dlx(Puzzle) ->
    build_dlx_matrix(Puzzle),
    cover_clues(Puzzle),

    Solution = array:new([{size, 81}, {default, -1}]),
    case dlx_search(root, 0, Solution) of
        {ok, Sol} -> {ok, extract_solution(Puzzle, Sol)};
        error -> error
    end.

%% Extract solution
extract_solution(Puzzle, Solution) ->
    %% Start with original puzzle
    FinalGrid = array:map(fun(R, RowArray) ->
        array:map(fun(C, Val) -> Val end, RowArray)
    end, Puzzle),

    %% Apply solution
    extract_solution_rows(0, Solution, FinalGrid).

extract_solution_rows(81, _, Grid) -> Grid;
extract_solution_rows(I, Solution, Grid) ->
    case array:get(I, Solution) of
        -1 -> extract_solution_rows(I + 1, Solution, Grid);
        RowId ->
            case ets:lookup(state, {row_info, RowId}) of
                [{{row_info, RowId}, {R, C, Num}}] ->
                    RowArray = array:get(R, Grid),
                    NewRowArray = array:set(C, Num, RowArray),
                    NewGrid = array:set(R, NewRowArray, Grid),
                    extract_solution_rows(I + 1, Solution, NewGrid);
                [] ->
                    extract_solution_rows(I + 1, Solution, Grid)
            end
    end.
